pytorch_lightning/utilities/data.py

```diff
@@ -336,23 +352,28 @@ def wrapper(obj: DataLoader, *args: Any, **kwargs: Any) -> None:
             if param.name != "self" and param.kind not in (param.VAR_POSITIONAL, param.VAR_KEYWORD)
         ]
 
-        cls = type(obj)
+        set_arg_names = []
         for arg_name, arg_value in chain(zip(param_names, args), kwargs.items()):
-            if hasattr(cls, arg_name) and getattr(cls, arg_name).fset is None:
-                # the class defines a read-only (no setter) property of this name. it's likely that the implementation
-                # will set `self._arg_name = arg_value` in `__init__` which is the attribute returned by the `arg_name`
-                # property so we are fine skipping in that case
-                continue
-            setattr(obj, arg_name, arg_value)
+            if not hasattr(obj, "__" + arg_name):
+                # Set the value privately if it has not been set yet. This achieves two things:
+                # 1. The argument is the one that was passed in the outermost call
+                #    and not overriden by a call to `super().__init__()`
+                # 2. The argument is the passed value and does not get overwritten in `__init__()`
+                setattr(obj, "__" + arg_name, arg_value)
+                set_arg_names.append(arg_name)
+        if not hasattr(obj, "_set_arg_names"):
+            obj._set_arg_names = set(set_arg_names)
+
         init(obj, *args, **kwargs)
 
     return wrapper
 
 
 # https://stackoverflow.com/a/63851681/9201239
 def _get_all_subclasses(cls: Type[Any]) -> Set[Type[Any]]:
-    """Returns a list of all classes that inherit directly or indirectly from the given class."""
-    subclasses = set()
+    """Returns a list of all classes that inherit directly or indirectly from the given class, including the base
+    class."""
+    subclasses = {cls}
```

Comment:
I would not make this modification. For example, it's used in the CLI where we don't want to include the class itself.

I suggest creating a new function that includes itself or manually adding it where you need it.

Reply:
So, code inspection found only one usage of this function in the code base and that is in the `_replace_dataloader_init_method`. That's why I thought it to be safe to change. However, I can absolutely change it back.